go/src/koding/klient/machine/mount/sync.go

```diff
@@ -0,0 +1,297 @@
+package mount
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"koding/klient/machine"
+	"koding/klient/machine/client"
+	"koding/klient/machine/index"
+
+	"github.com/koding/logging"
+)
+
+const (
+	LocalIndexName  = "index.local"  // file name of local directory index.
+	RemoteIndexName = "index.remote" // file name of remote directory index.
+)
+
+// DynamicClientFunc is an adapter that allows to dynamically provide clients
+// from a given mount ID. Error should be of ErrMountNotFound type when it is
+// not possible to find any client based on a given mount.
+type DynamicClientFunc func(ID) (client.Client, error)
+
+// Info stores information about current mount status.
+type Info struct {
+	ID    ID    // Mount ID.
+	Mount Mount // Mount paths stored in absolute form.
+
+	SyncCount int // Number of synced files.
+	AllCount  int // Number of all files handled by mount.
+
+	SyncDiskSize int64 // Total size of synced files.
+	AllDiskSize  int64 // Size of all files handled by mount.
+}
+
+// SyncOpts are the options used to configure Syncs object.
+type SyncOpts struct {
+	// ClientFunc is a factory for dynamic clients.
+	ClientFunc DynamicClientFunc
+
+	// WorkDir is a working directory that will be used by syncs object. The
+	// directory structure for single mount with ID will look like:
+	//
+	//   WorkDir
+	//   +-mount-<ID>
+	//     |-data
+	//     | +-... // mounted directory cache.
+	//     |-index.remote
+	//     +-index.local
+	//
+	WorkDir string
+
+	// Log is used for logging. If nil, default logger will be created.
+	Log logging.Logger
+}
+
+// Valid checks if provided options are correct.
+func (opts *SyncOpts) Valid() error {
+	if opts.ClientFunc == nil {
+		return errors.New("nil dynamic client function")
+	}
+
+	if opts.WorkDir == "" {
+		return errors.New("working directory is not set")
+	}
+
+	if _, err := os.Stat(opts.WorkDir); err != nil {
+		return fmt.Errorf("invalid working directory %q: %s", opts.WorkDir, err)
+	}
+
+	return nil
+}
+
+// Sync synchronizes and manages one or more directories. This type is
+// responsible for ensuring that remote and local directories are as much
+// similar as possible.
+type Sync struct {
+	opts SyncOpts
+	log  logging.Logger
+
+	mu    sync.RWMutex
+	syncs map[ID]*synced
+}
+
+// NewSync creates a new Sync instance from the given options.
+func NewSync(opts SyncOpts) (*Sync, error) {
+	if err := opts.Valid(); err != nil {
+		return nil, err
+	}
+
+	s := &Sync{
+		opts:  opts,
+		syncs: make(map[ID]*synced),
+	}
+
+	if opts.Log != nil {
+		s.log = opts.Log.New("sync")
+	} else {
+		s.log = machine.DefaultLogger.New("sync")
+	}
+
+	return s, nil
+}
+
+// Add starts synchronization between remote and local directories. It creates
+// all necessary cache files if they are not present.
+func (s *Sync) Add(id ID, m Mount) error {
+	s.mu.RLock()
+	sd, ok := s.syncs[id]
+	s.mu.RUnlock()
+
+	if ok {
```

Comment:
I'm not sure these fine-grained locks are correct. Two different threads can simultaneously pass this check and both can assign a new *synced value with the same id to the map. While if the whole body was protected with a single lock, the second thread would fail here.